cmd/server/helper_cert.go

```
@@ -57,66 +56,70 @@ func AttachCertificate(priv *jose.JSONWebKey, cert *x509.Certificate) {
 
-func getLock(set string) *sync.RWMutex {
-	mapLock.Lock()
-	defer mapLock.Unlock()
-	if _, ok := locks[set]; !ok {
-		locks[set] = new(sync.RWMutex)
-	}
-	return locks[set]
-}
-
-func GetOrCreateTLSCertificate(ctx context.Context, cmd *cobra.Command, d driver.Registry, iface config.ServeInterface) []tls.Certificate {
-	getLock(TlsKeyName).Lock()
-	defer getLock(TlsKeyName).Unlock()
-
-	cert, err := d.Config().TLS(ctx, iface).Certificate()
+// GetOrCreateTLSCertificate returns a function for use with
+// "net/tls".Config.GetCertificate. If the certificate and key are read from
+// disk, they will be automatically reloaded until stopReload is close()'d.
+func GetOrCreateTLSCertificate(ctx context.Context, d driver.Registry, iface config.ServeInterface, stopReload <-chan struct{}) func(*tls.ClientHelloInfo) (*tls.Certificate, error) {
+	lock.Lock()
+	defer lock.Unlock()
 
+	// check if certificates are configured
+	certFunc, err := d.Config().TLS(ctx, iface).GetCertificateFunc(stopReload, d.Logger())
 	if err == nil {
-		return cert
+		return certFunc
 	} else if !errors.Is(err, tlsx.ErrNoCertificatesConfigured) {
-		d.Logger().WithError(err).Fatalf("Unable to load HTTPS TLS Certificate")
+		d.Logger().WithError(err).Fatal("Unable to load HTTPS TLS Certificate")
+		return nil // in case Fatal is hooked
 	}
 
+	// no certificates configured: self-sign a new cert
 	priv, err := jwk.GetOrGenerateKeys(ctx, d, d.SoftwareKeyManager(), TlsKeyName, uuid.Must(uuid.NewV4()).String(), "RS256")
 	if err != nil {
 		d.Logger().WithError(err).Fatal("Unable to fetch or generate HTTPS TLS key pair")
+		return nil // in case Fatal is hooked
 	}
 
 	if len(priv.Certificates) == 0 {
 		cert, err := tlsx.CreateSelfSignedCertificate(priv.Key)
 		if err != nil {
-			d.Logger().WithError(err).Fatalf(`Could not generate a self signed TLS certificate`)
+			d.Logger().WithError(err).Fatal(`Could not generate a self signed TLS certificate`)
+			return nil // in case Fatal is hooked
```

I think we should instead return the error